modules/solid_mechanics/src/materials/StrainAdjustedDensity.C

```diff
@@ -0,0 +1,105 @@
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "StrainAdjustedDensity.h"
+
+registerMooseObject("SolidMechanicsApp", StrainAdjustedDensity);
+registerMooseObject("SolidMechanicsApp", ADStrainAdjustedDensity);
+
+template <bool is_ad>
+InputParameters
+StrainAdjustedDensityTempl<is_ad>::validParams()
+{
+  InputParameters params = Material::validParams();
+
+  params.addRequiredCoupledVar(
+      "displacements",
+      "The displacements appropriate for the simulation geometry and coordinate system. If "
+      "displacements are not coupled, a different material class such as GenericMaterialProperty "
+      "or ParsedMaterial should be used.");
+
+  params.addParam<std::string>("base_name",
+                               "Optional parameter that allows the user to define "
+                               "multiple material systems on the same block, "
+                               "e.g. for multiple phases");
+  params.addRequiredParam<MaterialPropertyName>("strain_free_density",
+                                                "Material property for strain-free density");
+  params.addClassDescription("Creates density material property");
+
+  return params;
+}
+
+template <bool is_ad>
+StrainAdjustedDensityTempl<is_ad>::StrainAdjustedDensityTempl(const InputParameters & parameters)
+  : Material(parameters),
+    _coord_system(getBlockCoordSystem()),
+    _disp_r(this->template coupledGenericValue<is_ad>("displacements", 0)),
+    _base_name(isParamValid("base_name") ? getParam<std::string>("base_name") + "_" : ""),
+    _strain_free_density(
+        this->template getGenericMaterialProperty<Real, is_ad>("strain_free_density")),
+    _grad_disp(this->template coupledGenericGradients<is_ad>("displacements")),
+    _density(declareGenericProperty<Real, is_ad>(_base_name + "density"))
+{
+  if (getParam<bool>("use_displaced_mesh"))
+    paramError("use_displaced_mesh",
+               "StrainAdjustedDensity needs to act on an undisplaced mesh. Use of a displaced mesh "
+               "leads to "
+               "incorrect gradient values");
+
+  // fill remaining components with zero
+  _grad_disp.resize(3, &genericZeroGradient<is_ad>());
+}
+
+template <bool is_ad>
+void
+StrainAdjustedDensityTempl<is_ad>::initQpStatefulProperties()
+{
+  _density[_qp] = _strain_free_density[_qp];
+}
+
+template <bool is_ad>
+void
+StrainAdjustedDensityTempl<is_ad>::computeQpProperties()
+{
+  _density[_qp] = _strain_free_density[_qp];
+
+  const auto Axx = (*_grad_disp[0])[_qp](0) + 1.0;
+  const auto & Axy = (*_grad_disp[0])[_qp](1);
+  const auto & Axz = (*_grad_disp[0])[_qp](2);
+  const auto & Ayx = (*_grad_disp[1])[_qp](0);
+  auto Ayy = (*_grad_disp[1])[_qp](1) + 1.0;
+  const auto & Ayz = (*_grad_disp[1])[_qp](2);
+  const auto & Azx = (*_grad_disp[2])[_qp](0);
+  const auto & Azy = (*_grad_disp[2])[_qp](1);
+  auto Azz = (*_grad_disp[2])[_qp](2) + 1.0;
```

Comment:
This is probably just copied but we might as well update it to use the names from rows (or columns?) constructor and the addIa (add identity) function. That would make this a two liner and more expressive at the same time.

Reply:
Can you help me with an example?

Reply:
yeah:

````
auto A = RankTwoTensor::initializeFromRows(*_grad_disp[0], *_grad_disp[1], *_grad_disp[2]);
A.addIa(1.0);
```

Reply:
oh, nice!